### PR TITLE
Ldamodel cython

### PR DESCRIPTION
**Pull request by TheWalkingDelirium**
Cython extension for speeding up LdaModel added. Added methods in the **lda_model.pyx**: _dirichlet_expectation, mean_ and _psi function_. Based on [this code from scikit_learn](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/decomposition/_online_lda.pyx)

@cscorley Hi. I run the [toy test](https://gist.github.com/tmylk/b71bf7d3ec2f203bfce2/5e9d9be47e6bb381109965fa0fe91e1de69496dd) and it always fail. Possibly because something has been changed in ldamodel method parameters in last commits. Could you try to run that test on your side?

### CODE REVIEW

**Comment by cscorley** [comment]
Thanks for getting Cython optimization started in LdaModel, @TheWalkingDelirium!

Right now I'm traveling, and it'll be until next weekend before I can look at this. Any more specific info about what it is failing on? Maybe @tmylk can have a look.

**Comment by tmylk** [comment]
Expected outcome:

-  3 heavily logged files of test_toy.py should agree:
   1) existing version
   2) pure python version
   3) cython version

- add tests to tests both pure python and cython code paths

- remove so files

File: gensim/models/ldamodel.py [hunk 1]
```diff
@@ -44,7 +44,27 @@
 from six.moves import xrange
 import six
 
-# log(sum(exp(x))) that tries to avoid overflow
+FAST_VERSION = 0
```
Comment (tmylk) [inline gensim/models/ldamodel.py]: Follow same approach as in word2vec: from gensim.models.word2vec_inner import FAST_VERSION

File: gensim/models/ldamodel.py [hunk 2]
```diff
@@ -99,6 +122,7 @@ class LdaState(utils.SaveLoad):
     reduce traffic.
 
     """
+
```
Comment (tmylk) [inline gensim/models/ldamodel.py]: remove the extra line between docstring and function

File: gensim/models/ldamodel.py [hunk 3]
```diff
@@ -198,6 +224,7 @@ class LdaModel(interfaces.TransformationABC):
 
     Model persistency is achieved through its `load`/`save` methods.
     """
+
```
Comment (tmylk) [inline gensim/models/ldamodel.py]: remove the extra line between docstring and function

File: gensim/models/ldamodel.py [hunk 4]
```diff
@@ -321,33 +351,33 @@ def __init__(self, corpus=None, num_topics=100, id2word=None,
...
-                init_prior = numpy.asarray([1.0 / (i + numpy.sqrt(self.num_topics)) for i in xrange(self.num_topics)])
-                init_prior /= init_prior.sum()
-                logger.info("using asymmetric %s %s", name, list(init_prior))
-            elif prior == 'auto':
-                is_auto = True
-                init_prior = numpy.asarray([1.0 / self.num_topics for i in xrange(self.num_topics)])
-                logger.info("using autotuned %s, starting with %s", name, list(init_prior))
-            else:
-                raise ValueError("Unable to determine proper %s value given '%s'" % (name, prior))
+        if prior == 'symmetric' or prior is None:
```
Comment (tmylk) [inline gensim/models/ldamodel.py]: what is the reason for this change? it is not necessary for the speed up

File: gensim/models/ldamodel.py [hunk 5]
```diff
@@ -420,11 +460,7 @@ def inference(self, chunk, collect_sstats=False):
         # Lee&Seung trick which speeds things up by an order of magnitude, compared
         # to Blei's original LDA-C code, cool!).
         for d, doc in enumerate(chunk):
-            if doc and not isinstance(doc[0][0], six.integer_types):
-                # make sure the term IDs are ints, otherwise numpy will get upset
-                ids = [int(id) for id, _ in doc]
-            else:
-                ids = [id for id, _ in doc]
+            ids = [id for id, _ in doc]
```
Comment (tmylk) [inline gensim/models/ldamodel.py]: what is the reason for this change? it has no relation to cython speed up

File: gensim/models/ldamodel.py [hunk 6]
```diff
@@ -518,18 +581,27 @@ def log_perplexity(self, chunk, total_docs=None):
...
+            bound_var = self.bound(chunk, subsample_ratio=subsample_ratio)
+            word_cnt = (subsample_ratio * corpus_words)
+            logger.debug("bound %s, word_cnt %s, perword_bound %s", bound_var, word_cnt, perwordbound)
+
         return perwordbound
 
     def update(self, corpus, chunksize=None, decay=None, offset=None,
                passes=None, update_every=None, eval_every=None, iterations=None,
-               gamma_threshold=None, chunks_as_numpy=False):
+               gamma_threshold=None):
```
Comment (tmylk) [inline gensim/models/ldamodel.py]: what is the reason for this change? it has no relation to cython speed up

File: gensim/models/ldamodel.py [hunk 7]
```diff
@@ -548,23 +620,6 @@ def update(self, corpus, chunksize=None, decay=None, offset=None,
...
-            into floats, these will be converted back into integers in
-            inference, which incurs a performance hit. For distributed
-            computing it may be desirable to keep the chunks as numpy
-            arrays.
-
-        See Also
-        --------
-        For other parameter settings see LdaModel().
-
         """
```
Comment (tmylk) [inline gensim/models/ldamodel.py]: what is the reason for this change? it has no relation to cython speed up

File: gensim/models/ldamodel.py [hunk 8]
```diff
@@ -632,7 +690,7 @@ def rho():
             dirty = False
 
             reallen = 0
-            for chunk_no, chunk in enumerate(utils.grouper(corpus, chunksize, as_numpy=chunks_as_numpy)):
+            for chunk_no, chunk in enumerate(utils.grouper(corpus, chunksize, as_numpy=True)):
```
Comment (tmylk) [inline gensim/models/ldamodel.py]: what is the reason for this change? it has no relation to cython speed up

File: gensim/models/ldamodel.py [hunk 9]
```diff
@@ -867,11 +951,10 @@ def top_topics(self, corpus, num_words=20):
             for m in top_words[1:]:
                 # m_docs is v_m^(t)
                 m_docs = doc_word_list[m]
-                m_index = numpy.where(top_words == m)[0]
```
Comment (tmylk) [inline gensim/models/ldamodel.py]: what is the reason for this change? it has no relation to cython speed up

File: gensim/models/ldamodel.py [hunk 10]
```diff
@@ -867,11 +951,10 @@ def top_topics(self, corpus, num_words=20):
...
 
                 # Sum of top words l=1..m-1
                 # i.e., all words ranked higher than the current word m
-                for l in top_words[:m_index - 1]:
+                for l in top_words[:m - 1]:
```
Comment (tmylk) [inline gensim/models/ldamodel.py]: what is the reason for this change? it has no relation to cython speed up

File: gensim/models/ldamodel.py [hunk 11]
```diff
@@ -898,7 +981,6 @@ def get_document_topics(self, bow, minimum_probability=None):
         """
         if minimum_probability is None:
             minimum_probability = self.minimum_probability
-        minimum_probability = max(abs(minimum_probability), 1e-8)  # never allow zero values in sparse output
```
Comment (tmylk) [inline gensim/models/ldamodel.py]: what is the reason for this change? it has no relation to cython speed up

**Comment by tmylk** [comment]
Thanks!

Outstanding items:

- Remove all changes from this PR that are outside of cython code

- Tests for the cython and non-cython code

- Performance test results on [20 newsgroups](https://gist.github.com/tmylk/7e934806715bcb56a78f)

- updated CHANGELOG.txt

File: gensim/models/ldamodel.py [hunk 12]
```diff
@@ -282,11 +310,13 @@ def __init__(self, corpus=None, num_topics=100, id2word=None,
         self.update_every = update_every
         self.eval_every = eval_every
 
-        self.alpha, self.optimize_alpha = self.init_dir_prior(alpha, 'alpha')
+        self.optimize_alpha = alpha == 'auto'
+        self.alpha = self.init_dir_prior(alpha, 'alpha')
```
Comment (tmylk) [inline gensim/models/ldamodel.py]: what is the reason for this change? it is not related to cython

File: gensim/models/ldamodel.py [hunk 13]
```diff
@@ -282,11 +310,13 @@ def __init__(self, corpus=None, num_topics=100, id2word=None,
...
 
         assert self.alpha.shape == (num_topics,), "Invalid alpha shape. Got shape %s, but expected (%d, )" % (str(self.alpha.shape), num_topics)
 
-        self.eta, self.optimize_eta = self.init_dir_prior(eta, 'eta')
+        self.optimize_eta = eta == 'auto'
```
Comment (tmylk) [inline gensim/models/ldamodel.py]: what is the reason for this change? it is not related to cython

File: gensim/models/ldamodel.py [hunk 14]
```diff
@@ -321,33 +351,33 @@ def __init__(self, corpus=None, num_topics=100, id2word=None,
...
+        # !!! first change by Lev
         self.expElogbeta = numpy.exp(dirichlet_expectation(self.state.sstats))
+        if __debug__:
+            logger.debug("Initial expElogbeta %s", str(self.expElogbeta))
 
         # if a training corpus was provided, start estimating the model right away
         if corpus is not None:
-            use_numpy = self.dispatcher is not None
-            self.update(corpus, chunks_as_numpy=use_numpy)
+            self.update(corpus)
```
Comment (tmylk) [inline gensim/models/ldamodel.py]: what is the reason for this change? it is not related to cython

File: gensim/models/ldamodel.py [hunk 15]
```diff
@@ -548,23 +620,6 @@ def update(self, corpus, chunksize=None, decay=None, offset=None,
...
-            array of not. Numpy can in some settings turn the term IDs
-            into floats, these will be converted back into integers in
-            inference, which incurs a performance hit. For distributed
-            computing it may be desirable to keep the chunks as numpy
-            arrays.
-
-        See Also
-        --------
-        For other parameter settings see LdaModel().
-
```
Comment (tmylk) [inline gensim/models/ldamodel.py]: what is the reason for this change? it is not related to cython

File: gensim/models/ldamodel.py [hunk 16]
```diff
@@ -632,7 +690,7 @@ def rho():
             dirty = False
 
             reallen = 0
-            for chunk_no, chunk in enumerate(utils.grouper(corpus, chunksize, as_numpy=chunks_as_numpy)):
```
Comment (tmylk) [inline gensim/models/ldamodel.py]: what is the reason for this change? it is not related to cython

File: gensim/models/ldamodel.py (hunk 9 above)
Comment (tmylk) [inline gensim/models/ldamodel.py]: what is the reason for this change? it is not related to cython

**Comment by TheWalkingDelirium** [comment]
https://gist.github.com/TheWalkingDelirium/7bc7e867050b192dd6d2 
test results and profiling screenshot

**Comment by menshikh-iv** [comment]
Ping @TheWalkingDelirium, what status of this PR? Will you finish it soon?

**Comment by TheWalkingDelirium** [comment]
Never, sorry.

**Comment by piskvorky** [comment]
@menshikh-iv what was actually missing here? Looks like a very cool change and @TheWalkingDelirium answered @tmylk 's points above?
